[[ FFI ]] Tidy up libscript FFI support

runrevmark opened the pull request:
This patch tidies up the current implementation of FFI in libscript.

The type of foreign handler is now determined by the 'language'
field. This is set after successful parsing of the binding string and
can be one of 'C', 'BuiltinC' or 'Java'.

Parsing of the binding string has been split up into per-language
functions. The language prefix of the string is first parsed out, and
that is used to determine which further function to call.

The C++ and ObjC related errors and bindings have been removed as they
are both to be implemented via C bindings and therefore the distinction
is unnecessary.

runrevmark committed:
[[ FFI ]] Tidy up libscript FFI support

This patch tidies up the current implementation of FFI in libscript.

The type of foreign handler is now determined by the 'language'
field. This is set after successful parsing of the binding string and
can be one of 'C', 'BuiltinC' or 'Java'.

Parsing of the binding string has been split up into per-language
functions. The language prefix of the string is first parsed out, and
that is used to determine which further function to call.

The C++ and ObjC related errors and bindings have been removed as they
are both to be implemented via C bindings and therefore the distinction
is unnecessary.

livecodeali commented:
@livecode-vulcan review ok d0175bf

livecode-vulcan commented:
💙 review by @livecodeali ok d0175bfb87f0e5ea2bce640d06951ccfdb62a262

livecode-vulcan commented:
😞 test failure d0175bfb87f0e5ea2bce640d06951ccfdb62a262

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/435)
* try-community-armv6-android-api9: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/443)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/422)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/365)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/365)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/430)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/435)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/433)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/440)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/433)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/440)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/459)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/429)

livecodeali commented on libscript/src/script-execute.cpp:
```diff
@@ -167,16 +167,17 @@ class MCScriptForeignInvocation
+        case kMCScriptForeignHandlerLanguageJava:
 #ifdef TARGET_SUBPLATFORM_ANDROID
             extern bool MCAndroidIsOnSystemThread();
             if (!p_handler->is_ui_bound || MCAndroidIsOnSystemThread())
```
From buildbot log, this should be `p_handler->java.is_ui_bound`

runrevmark committed:
[[ FFI ]] Fix incorrect reference to is_ui_bound

This patch fixes the missing 'java' part when unpacking the
is_ui_bound java-specific foreign handler var.

livecodeali commented:
@livecode-vulcan review ok 02e2dfc

runrevmark commented:
@livecode-vulcan review by @livecodeali ok 10da49

livecode-vulcan commented:
💙 review by @livecodeali ok 10da49c82ab9c4ad7e47d150d0bfcd498a5f6570

livecode-vulcan commented:
😞 test failure 10da49c82ab9c4ad7e47d150d0bfcd498a5f6570

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/442)
* try-community-armv6-android-api9: [pending](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/450)
* try-community-js-emscripten-sdk1.35: [pending](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/429)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/372)
* try-community-universal-ios-iphonesimulator10.3: [pending](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/372)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/437)
* try-community-universal-mac-macosx10.9: [pending](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/442)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/440)
* try-community-x86-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/449)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/440)
* try-community-x86_64-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/449)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/467)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/436)

livecodeali commented:
@livecode-vulcan retry 10da49c

livecode-vulcan commented:
😎 test success 10da49c82ab9c4ad7e47d150d0bfcd498a5f6570

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/443)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/451)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/430)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/373)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/373)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/438)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/443)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/441)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/451)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/441)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/450)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/468)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/437)
